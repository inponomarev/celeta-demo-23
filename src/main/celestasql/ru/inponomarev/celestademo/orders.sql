create schema orders version '1.0';

create sequence customer_id;

create table customer (
    id int not null default nextval(customer_id) primary key,
    name varchar(50) not null,
    email varchar(50),
    description varchar(200)
);

create index customer_idx on customer(name);

create table item (
    id varchar(12) not null primary key,
    name varchar(100) not null,
    default_price int
);

create sequence order_id;

create table order (
    id int not null default nextval(order_id) primary key,
    customer_id int not null foreign key references customer(id),
    item_id varchar(12) not null foreign key references item(id),
    quantity int not null default 0,
    price int not null default 0,
    amount int not null default 0
);

create materialized view item_orders as
    select
        item_id,
        sum(quantity) as ordered_quantity,
        sum(amount) as ordered_amount
    from order group by item_id;

create view item_view as
    select
      i.id as id,
      i.name as name,
      o.ordered_quantity as ordered_quantity,
      o.ordered_amount as ordered_amount
    from item as i left join item_orders as o on i.id = o.item_id;
