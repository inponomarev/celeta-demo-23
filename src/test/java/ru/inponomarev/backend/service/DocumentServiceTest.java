package ru.inponomarev.backend.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.approvaltests.Approvals;
import org.approvaltests.JsonApprovals;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ru.curs.celesta.CallContext;
import ru.curs.celestaunit.CelestaTest;
import ru.inponomarev.backend.dto.ItemDTO;
import ru.inponomarev.backend.dto.OrderDTO;
import ru.inponomarev.celestademo.CustomerCursor;
import ru.inponomarev.celestademo.ItemCursor;
import ru.inponomarev.celestademo.OrderCursor;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@CelestaTest
class DocumentServiceTest {

    private final OrderService orderService = new OrderService();
    private CustomerCursor customer;
    private ItemCursor item;

    @BeforeEach
    void setUp(CallContext ctx) {
        customer = new CustomerCursor(ctx);
        customer.setName("John Doe")
                .setEmail("john@example.com")
                .insert();
        item = new ItemCursor(ctx);
        item.setId("12345")
                .setName("cheese")
                .setDefaultPrice(42)
                .insert();
    }

    @Test
    void orderIsPostedWithDefaultPrice(CallContext ctx) throws JsonProcessingException {
        //ARRANGE
        OrderDTO dto =  new OrderDTO();
                dto.setCustomerId(customer.getId());
                dto.setItemId(item.getId());
                dto.setQuantity(100);

        //ACT
        OrderDTO result = orderService.postOrder(ctx, dto);

        //ASSERT
        OrderCursor orderCursor = new OrderCursor(ctx);
        Assertions.assertThat(orderCursor.count()).isEqualTo(1);
        orderCursor.first();
        Assertions.assertThat(orderCursor.getPrice()).isEqualTo(item.getDefaultPrice());
        Assertions.assertThat(orderCursor.getQuantity()).isEqualTo(100);
        Assertions.assertThat(orderCursor.getAmount()).isEqualTo(item.getDefaultPrice() * orderCursor.getQuantity());
        JsonApprovals.verifyJson(new ObjectMapper().writer().writeValueAsString(result));
    }

    @Test
    void orderedItemsMethodReturnsAggregatedValues(CallContext ctx) throws JsonProcessingException {
        //ARRANGE
        ItemCursor item2 = new ItemCursor(ctx);
        item2.setId("2").setName("item 2").insert();
        OrderCursor orderCursor = new OrderCursor(ctx);
        orderCursor.setId(null).setItemId(item.getId())
                .setCustomerId(customer.getId()).setQuantity(1).setAmount(2).insert();
        orderCursor.setId(null).setItemId(item.getId())
                .setCustomerId(customer.getId()).setQuantity(3).setAmount(3).insert();
        orderCursor.setId(null).setItemId(item2.getId())
                .setCustomerId(customer.getId()).setQuantity(5).setAmount(4).insert();
        //ACT
        List<ItemDTO> result = orderService.getAggregateReport(ctx);

        //ASSERT
        JsonApprovals.verifyJson(new ObjectMapper().writer().writeValueAsString(result));
    }
}