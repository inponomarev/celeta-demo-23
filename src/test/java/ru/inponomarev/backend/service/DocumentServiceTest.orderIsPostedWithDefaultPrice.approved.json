{
  "id": 1,
  "customer_id": 1,
  "item_id": "12345",
  "quantity": 100,
  "price": 42,
  "amount": 4200
}